File: data-structures/singly-linked-list/reverse.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll                  long long int
#define vi                  vector <long long int>
#define inf                 1e18
#define mod                 1000000007
#define loop(i,a,b)         for(int i=(a);i<=(b);i++)
#define pb                  push_back
#define vs                  vector <string>

struct node{
    int data;
    struct node *next;
};
//Reverse Linked List
struct node *reverse(struct node *head){
    struct node *prev=NULL,*curr=head,*next=NULL;
    if(head==NULL){
        cout<<"Empty Linked List";
        return NULL;
    }
    while(curr!=NULL){
        next=curr->next;
        curr->next=prev;
        prev=curr;
        curr=next;
    }
    return head=prev;


}

void traverse(struct node * head){
    cout<<endl<<"Element of LL :";
    struct node *curr=head;
    while(curr!=NULL){
        cout<<curr->data<<" ";
        curr=curr->next;
    }
}

struct node * insertAtLast(struct node * head,int data){
    struct node *curr=head,*newNode=(struct node *)malloc(sizeof(struct node ));
    newNode->data=data;
    newNode->next=NULL;
    if(head==NULL){
        return head=newNode;
    }else{
        while(curr->next!=NULL){
            curr=curr->next;
        }
        curr->next=newNode;
        return head;
    }
}
int main(){
    int n,data;
    struct node *head=NULL;
    cin>>n;
    loop(i,0,n-1){
        cin>>data;
       head=insertAtLast(head,data);
    }
    traverse(head);
    cout<<"After Reversal";
    head=reverse(head);
    traverse(head);
    return 0;
}

File: data-structures/singly-linked-list/reverse.py
class Node:
    def __init__(self,val):
        self.data=val
        self.next=None

        

class LL:
    def __init__(self):
        self.head=None

    def insert(self,data):
        newNode=Node(data)
        if self.head is None:
            self.head=newNode
            return
        curr=self.head
        while(curr.next):
            curr=curr.next
        curr.next=newNode        

    def traverse(self):
        temp=self.head
        if temp is None:
            print("Empty List",sep=" ")
            return
        while(temp):
            print(temp.data,sep=" ")
            temp=temp.next

    def reverse(self):
        prev=None
        nxt=None
        curr=self.head

        while(curr):
            nxt=curr.next
            curr.next=prev 
            prev=curr
            curr=nxt
        self.head=prev           
if __name__ =='__main__':
    ll=LL()
    ll.insert(1)
    ll.insert(2)
    ll.insert(3)

    ll.traverse()
    ll.reverse()
    ll.traverse()
    
